ipa-core/src/protocol/ipa_prf/validation_protocol/proof_generation.rs

```diff
@@ -165,52 +189,40 @@ impl ProofBatch {
     ///
     /// ## Errors
     /// Propagates error from sending values over the network channel.
-    pub async fn send_to_left<C>(&self, ctx: &C) -> Result<(), Error>
+    pub async fn send_to_left<C>(&self, ctx: &C, record_id: RecordId) -> Result<(), Error>
     where
         C: Context,
     {
-        // set up context for the communication over the network
-        let communication_ctx = ctx.set_total_records(TotalRecords::specified(self.len())?);
-
-        // set up channel
-        let send_channel_left =
-            &communication_ctx.send_channel::<Fp61BitPrime>(ctx.role().peer(Direction::Left));
-
-        // send to left
-        // we send the proof batch via sending the individual field elements
-        communication_ctx
-            .parallel_join(
-                self.iter().enumerate().map(|(i, x)| async move {
-                    send_channel_left.send(RecordId::from(i), x).await
-                }),
-            )
-            .await?;
-        Ok(())
+        Ok(ctx
+            .set_total_records(TotalRecords::Indeterminate)
+            .send_channel::<Box<Array>>(ctx.role().peer(Direction::Left))
+            .send(record_id, self.to_array())
+            .await?)
     }
 
     /// This function receives a `Proof` from the party on the right.
     ///
     /// ## Errors
     /// Propagates errors from receiving values over the network channel.
-    pub async fn receive_from_right<C>(ctx: &C, length: usize) -> Result<Self, Error>
+    ///
+    /// ## Panics
+    /// If the recursion depth implied by `length` exceeds `MAX_PROOF_RECURSION`.
+    pub async fn receive_from_right<C>(
+        ctx: &C,
+        record_id: RecordId,
+        length: usize,
+    ) -> Result<Self, Error>
     where
         C: Context,
     {
-        // set up context
-        let communication_ctx = ctx.set_total_records(TotalRecords::specified(length)?);
-
-        // set up channel
-        let receive_channel_right =
-            &communication_ctx.recv_channel::<Fp61BitPrime>(ctx.role().peer(Direction::Right));
-
-        // receive from the right
+        assert!(length <= ARRAY_LEN);
         Ok(ctx
-            .parallel_join(
-                (0..length)
-                    .map(|i| async move { receive_channel_right.receive(RecordId::from(i)).await }),
-            )
+            .set_total_records(TotalRecords::Indeterminate)
```

Comment:
oh, this is much better than parallel join, can't wait until I see the impact of this change

Reply:
I don't expect that this is big enough contributor that the difference will show up in overall runtime, but I do think that getting rid of the parallel join more than offsets any cost of space overhead in the fixed-size structures.